yt_dlp/extractor/educast.py

```diff
@@ -0,0 +1,455 @@
+    _VALID_URL = r'https?://(?:www\.)?educast\.fccn\.pt/vod/channels/(?P<id>[a-zA-Z0-9]+)/?(?:$|[?#])'
+    _TESTS = [
+        {
+            'note': 'test for private Educast Channel',
+            'url': 'https://educast.fccn.pt/vod/channels/2o0eonmrak',
+            'info_dict':
+            {
+                'id': '2o0eonmrak',
+                'title': 'Vídeos Institucionais FCT-FCCN',
+                'description': str,
+            },
+            'playlist_mincount': 26,
+        },
+        {
+            'note': 'test for private Educast Channel',
+            'url': 'https://educast.fccn.pt/vod/channels/2fudccnyj7',
+            'info_dict': {
+                'id': '2fudccnyj7',
+                'title': 'Cálculo Diferencial e Integral III - Aulas de Recuperação',
+                'description': str,
+            },
+            'playlist_mincount': 26,
+            'skip': 'This channel is private and requires authentication to access',
+        },
+    ]
+
+    def _extract_video_links_from_html(self, webpage, ie_key):
+        import re
+        videos_by_id = {}
+        pattern = r'href="https://educast\.fccn\.pt/vod/clips/(?P<id>[a-zA-Z0-9]+)/(?P<option>[^?"/]+)'
+        for m in re.finditer(pattern, webpage or '', re.IGNORECASE):
+            video_id = m.group('id')
+            option = m.group('option')
+            if video_id not in videos_by_id:
+                videos_by_id[video_id] = []
+            videos_by_id[video_id].append(option)
+
+        videos = []
+        for video_id, candidates in videos_by_id.items():
+            # prefer 'streaming.html'
+            candidates.sort(key=lambda x: x[0] == 'streaming.html')
+            chosen_url = f'{EducastIE._API_BASE}/vod/clips/{video_id}/{candidates[0]}'
+            videos.append({
+                '_type': 'url',
+                'url': chosen_url,
+                'ie_key': ie_key,
+                'id': video_id,
+            })
+        return videos
+
+    def _extract_videos(self, url, channel_id, webpage=None):
+        max_downloads = None
+        if hasattr(self, '_downloader') and self._downloader:
+            max_downloads = self._downloader.params.get('max_downloads')
+
+        def get_page(page):
+            import urllib.parse
+            url_parts = list(urllib.parse.urlparse(url))
+            query = urllib.parse.parse_qs(url_parts[4])
+            query['page'] = [str(page)]
+            url_parts[4] = urllib.parse.urlencode(query, doseq=True)
+            page_url = urllib.parse.urlunparse(url_parts)
+
+            return self._download_webpage(page_url, channel_id, note=f'Downloading page {page}', fatal=False)
+
+        def parse_func(page_result):
+            return self._extract_video_links_from_html(page_result, EducastIE.ie_key())
+
+        try:
+            videos = EducastIE._paginate_and_collect(get_page, parse_func, max_videos=max_downloads)
+            if videos:
+                return videos
```

Comment:
this is overly complicated and inefficient, but I'm not going to review the entries logic until after the other comments are addressed